src/drivers/imu/mpu9250/MPU9250_mag.cpp

```diff
@@ -347,3 +327,107 @@ MPU9250_mag::ak8963_setup()
 
 	return OK;
 }
+
+void MPU9250_mag::write_imu_reg_verified(int reg, uint8_t val, uint8_t mask)
+{
+	uint8_t b;
+	int retry = 5;
+
+	while (retry) { // should not reach any retries in normal condition
+		--retry;
+		_parent->write_reg(reg, val);
+
+		b = _parent->read_reg(reg);
+
+		if ((b & mask) != val) {
+			PX4_DEBUG("MPU9250_mag::write_imu_reg_verified failed. retrying...");
+			continue;
+
+		} else {
+			return;
+		}
+	}
+}
+
+void MPU9250_mag::read_reg_through_mpu9250(uint8_t reg, uint8_t *val)
+{
+	uint8_t b = 0;
+
+	// Read operation on the mag using the slave 4 registers.
+	write_imu_reg_verified(MPUREG_I2C_SLV4_ADDR,
+			       AK8963_I2C_ADDR | BIT_I2C_READ_FLAG, 0xff);
+
+	// Set the mag register to read from.
+	write_imu_reg_verified(MPUREG_I2C_SLV4_REG, reg, 0xff);
+
+	// Read the existing value of the SLV4 control register.
+	b = _parent->read_reg(MPUREG_I2C_SLV4_CTRL);
+
+	// Set the I2C_SLV4_EN bit in I2C_SL4_CTRL register without overwriting other
+	// bits. Enable data transfer, a read transfer as configured above.
+	b |= 0x80;
+	// Trigger the data transfer
+	_parent->write_reg(MPUREG_I2C_SLV4_CTRL, b);
+
+	// Continuously check I2C_MST_STATUS register value for the completion
+	// of I2C transfer until timeout
+
+	int loop_ctrl = 1000; // wait up to 1000 * 1ms for completion
+
+	do {
+		px4_usleep(1000);
```

Comment:
Some of these relatively long sleeps might be problematic when the bus is shared. For example on the pixracer (fmu-v4) the 2 IMUs (icm20602 + mpu9250) are on the same SPI and running on the same thread.